datastore/redis.go

```diff
@@ -294,3 +323,55 @@ func (r *RedisCache) GetBlockBuilderStatus(builderPubkey string) (isHighPrio, is
 	isBlacklisted = BlockBuilderStatus(res) == RedisBlockBuilderStatusBlacklisted
 	return isHighPrio, isBlacklisted, err
 }
+
+// SaveLatestBuilderBid saves the latest bid by a specific builder
+func (r *RedisCache) SaveLatestBuilderBid(slot uint64, builderPubkey, parentHash, proposerPubkey string, headerResp *types.GetHeaderResponse) (err error) {
+	keyLatestBids := r.keyBlockBuilderLatestBids(slot, parentHash, proposerPubkey)
+	err = r.HSetObj(keyLatestBids, builderPubkey, headerResp, expiryBidCache)
+	if err != nil {
+		return err
+	}
+
+	keyLatestBidsValue := r.keyBlockBuilderLatestBidsValue(slot, parentHash, proposerPubkey)
+	err = r.client.HSet(context.Background(), keyLatestBidsValue, builderPubkey, headerResp.Data.Message.Value.String()).Err()
+	if err != nil {
+		return err
+	}
+	return r.client.Expire(context.Background(), keyLatestBidsValue, expiryBidCache).Err()
+}
+
+func (r *RedisCache) UpdateTopBid(slot uint64, parentHash, proposerPubkey string) (err error) {
+	// Get all builder's latest submission values
+	keyBidValues := r.keyBlockBuilderLatestBidsValue(slot, parentHash, proposerPubkey)
+	bidValueMap, err := r.client.HGetAll(context.Background(), keyBidValues).Result()
+	if err != nil {
+		return err
+	}
+
+	// Find bid with highest value among all the latest bids
+	topBidValue := big.NewInt(0)
+	topBidBuilderPubkey := ""
+	for builderPubkey, bidValue := range bidValueMap {
+		val := new(big.Int)
+		val.SetString(bidValue, 10)
+		if val.Cmp(topBidValue) > 0 {
+			topBidValue = val
+			topBidBuilderPubkey = builderPubkey
+		}
+	}
+
+	if topBidBuilderPubkey == "" {
+		return ErrFailedUpdatingTopBidNoBids
```

Comment:
Should this result in no updates? To be safe you could clear (delete) the top bid for example.

Reply:
can't update because there are 0 bids available. this should never  happen, but needs to be checked anyway